firedrake/assemble.py

```diff
@@ -169,239 +160,240 @@ def thunk():
-              appctx={},
-              options_prefix=None,
-              assemble_now=False,
-              allocate_only=False,
-              zero_tensor=True,
-              diagonal=False):
-    r"""Assemble the form or Slate expression f and return a Firedrake object
-    representing the result. This will be a :class:`float` for 0-forms/rank-0
-    Slate tensors, a :class:`.Function` for 1-forms/rank-1 Slate tensors and
-    a :class:`.Matrix` for 2-forms/rank-2 Slate tensors.
+def get_matrix(form, mat_type, sub_mat_type, *, bcs=None,
+               options_prefix=None, tensor=None, appctx=None,
+               form_compiler_parameters=None):
+    matfree = mat_type == "matfree"
+    arguments = form.arguments()
+    if tensor is not None and tensor.a.arguments() != arguments:
+        raise ValueError("Form's arguments do not match provided result tensor")
+    if matfree:
+        if tensor is None:
+            tensor = matrix.ImplicitMatrix(form, bcs,
+                                           fc_params=form_compiler_parameters,
+                                           appctx=appctx,
+                                           options_prefix=options_prefix)
+        elif not isinstance(tensor, matrix.ImplicitMatrix):
+            raise ValueError("Expecting implicit matrix with matfree")
+        else:
+            pass
+        return tensor, (), lambda: tensor
 
-    :arg bcs: A tuple of :class`.DirichletBC`\s and/or :class`.EquationBCSplit`\s to be applied.
-    :arg tensor: An existing tensor object into which the form should be
-        assembled. If this is not supplied, a new tensor will be created for
-        the purpose.
-    :arg form_compiler_parameters: (optional) dict of parameters to pass to
-        the form compiler.
-    :arg inverse: (optional) if f is a 2-form, then assemble the inverse
-         of the local matrices.
-    :arg mat_type: (optional) type for assembled matrices, one of
-        "nest", "aij", "baij", or "matfree".
-    :arg sub_mat_type: (optional) type for assembled sub matrices
-        inside a "nest" matrix.  One of "aij" or "baij".
-    :arg appctx: Additional information to hang on the assembled
-         matrix if an implicit matrix is requested (mat_type "matfree").
-    :arg options_prefix: An options prefix for the PETSc matrix
-        (ignored if not assembling a bilinear form).
-    """
-    if mat_type is None:
-        mat_type = parameters.parameters["default_matrix_type"]
-    if mat_type not in ["matfree", "aij", "baij", "nest", "dense"]:
-        raise ValueError("Unrecognised matrix type, '%s'" % mat_type)
-    if sub_mat_type is None:
-        sub_mat_type = parameters.parameters["default_sub_matrix_type"]
-    if sub_mat_type not in ["aij", "baij"]:
-        raise ValueError("Invalid submatrix type, '%s' (not 'aij' or 'baij')", sub_mat_type)
+    if isinstance(form, slate.TensorBase):
+        # FIXME: inherit from slate form somehow.
+        integral_types = {"cell"}
```

Comment:
In the KernelBuilder of slac, the integral type is inherited from TSFC and TSFC inherits it from the UFL form if I understand it, right. Maybe something like `[integral.integral_type for integral in form.form.integrals()]` is what you need here?

Reply:
Well, a slate kernel only iterates over cells, so it can only make contributions through cell_node_maps (hence `cell` here). This is being used to determine the sparsity pattern of the resulting global matrix, so getting the integral types from the form inside the kernel is probably wrong.

I guess a slate `TensorBase` can advertise an `integrals()` method that returns a 1-iterable of objects that return `integral_type() == cell`.

Something like

```python
class MockIntegral(object):
    integral_type = lambda: "cell"

MockIntegral = MockIntegral() # singleton
...
class Tensor(...):
    integrals = (MockIntegral, )
```

Reply:
@wence- is correct, while a Slate expression may (and often does) contain integrals of _each_ type, it only makes sense to iterate cell-wise. It feels odd for a Slate expression to advertise an integral type.

An alternative would be to create a function that dispatches on the expression (UFL.form vs Slate.TensorBase). Something like:

```python
@singledispatch
def iteration_domains(expr, bcs):
    raise ValueError()

@iteration_domains.register(slate.TensorBase)
def _(expr, bcs):
    return {"cell"}

@iteration_domains.register(ufl.Form)
def _(expr, bcs):
    integral_types = set(i.integral_type() for i in expr.integrals())
    for bc in bcs:
        integral_types.update(integral.integral_type()
                              for integral in bc.integrals())
    return integral_types
```
Doing so would allow you to replace all this with
```python
integral_types = iteration_domains(expr, bcs)
```
This is just an alternative way to write this. I think it might be worth a discussion / debate whether or not Slate expressions should support something other than "cellwise" assembly. I personally think it should _always_ be cellwise.

I am willing to be convinced otherwise if there is a concrete application for assembling a Slate expression "facet-wise."

Reply:
Well, I think I would prefer just mocking the interface that ufl Forms provide and then there doesn't need to be any type-based dispatch here. The original idea was that slate objects would look "enough-like" forms that assembly didn't need to care. We didn't quite get there, but we nearly have!

Reply:
I did the mocking I sketched.